### PR TITLE
1st version Image Segmenter

### PR DESCRIPTION
**Pull request by ECEMACHINE**
### Which issue(s) does this Pull Request fix?

resolves #000

### CODE REVIEW

File: autokeras/tasks/image.py [hunk 1]
```diff
@@ -221,3 +221,103 @@ def fit(self,
...
+
+
+class ImageSegmenter(SupervisedImagePipeline):
+    """AutoKeras image segmentation class.
+
+    # Arguments
+        num_classes: Int. Defaults to None. If None, it will infer from the data.
+        loss: A loss function. Defaults to use 'tf.losses.softmax_cross_entropy'.
+        metrics: A list of metrics used to measure the accuracy of the model,
+        default to 'mIoU'
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: 1. Add an indent here.
2. a period in the end.
3. Let's use accuracy as default value.

File: autokeras/tasks/image.py [hunk 2]
```diff
@@ -221,3 +221,103 @@ def fit(self,
...
                     validation_data=validation_data,
                     **kwargs)
+
+
+class ImageSegmenter(SupervisedImagePipeline):
+    """AutoKeras image segmentation class.
+
+    # Arguments
+        num_classes: Int. Defaults to None. If None, it will infer from the data.
+        loss: A loss function. Defaults to use 'tf.losses.softmax_cross_entropy'.
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: We should use the same description as the image classifier.

File: autokeras/tasks/image.py [hunk 3]
```diff
@@ -221,3 +221,104 @@ def fit(self,
...
+
+        It will search for the best model based on the performances on
+        validation data.
+
+        # Arguments
+            x: numpy.ndarray or tensorflow.Dataset. Training image dataset x.
+                The shape of the data should be 3 or 4 dimensional, the last
+                dimension of which should be channel dimension.
+            y: numpy.ndarray or tensorflow.Dataset. Training image data
+                set y. It should be the groundtruth image data set as the label.
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: Change "It should be the groundtruth image data set as the label." to "It should be a tensor with the same shape of x. Each element in the tensor is the label of the corresponding pixel."

Comment (haifeng-jin) [reply autokeras/tasks/image.py]: https://github.com/keras-team/autokeras/compare/33879915c9a2ef9c1348170149c35dc13773fd41..ad552e780c02799cb3b64059ba08a278ce55e5f3#diff-173d8f18374c208cb0df47eb0bc0e6dbL266

Comment (haifeng-jin) [reply autokeras/tasks/image.py]: https://github.com/keras-team/autokeras/compare/33879915c9a2ef9c1348170149c35dc13773fd41..ad552e780c02799cb3b64059ba08a278ce55e5f3#diff-173d8f18374c208cb0df47eb0bc0e6dbL288-L289

Comment (haifeng-jin) [reply autokeras/tasks/image.py]: https://github.com/keras-team/autokeras/compare/33879915c9a2ef9c1348170149c35dc13773fd41..ad552e780c02799cb3b64059ba08a278ce55e5f3#diff-173d8f18374c208cb0df47eb0bc0e6dbL227

Comment (haifeng-jin) [reply autokeras/tasks/image.py]: These lines need to be changed back to it was before the force push. Try not to force-push when multiple people have committed to the same branch.

File: autokeras/tasks/image.py [hunk 4]
```diff
@@ -221,3 +221,107 @@ def fit(self,
...
+        super().__init__(
+            outputs=hypermodels.SegmenterHead(num_classes=num_classes,
+                                              multi_label=multi_label,
+                                              loss=loss,
+                                              metrics=metrics),
+            max_trials=max_trials,
+            directory=directory,
+            name=name,
+            objective=objective,
+            tuner=task_specific.ImageSegmenterTuner,
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: Let's use `greedy.Greedy` for now.

File: autokeras/tasks/image.py [hunk 5]
```diff
@@ -221,3 +221,107 @@ def fit(self,
                     validation_split=validation_split,
                     validation_data=validation_data,
                     **kwargs)
+
+
+class ImageSegmenter(SupervisedImagePipeline):
+    """AutoKeras image segmentation class.
+
+    # Arguments
+        num_classes: Int. Defaults to None. If None, it will infer from the data.
```
Comment (fchollet) [inline autokeras/tasks/image.py]: "it will be inferred from" (you can also apply this fix to the other docstrings)

File: autokeras/tasks/image.py [hunk 6]
```diff
@@ -221,3 +221,107 @@ def fit(self,
                     validation_split=validation_split,
                     validation_data=validation_data,
                     **kwargs)
+
+
+class ImageSegmenter(SupervisedImagePipeline):
+    """AutoKeras image segmentation class.
+
+    # Arguments
```
Comment (fchollet) [inline autokeras/tasks/image.py]: Consider including type annotations.

File: autokeras/tasks/image.py [hunk 7]
```diff
@@ -221,3 +221,107 @@ def fit(self,
...
+            **kwargs):
+        """Search for the best model and hyperparameters for the AutoModel.
+
+        It will search for the best model based on the performances on
+        validation data.
+
+        # Arguments
+            x: numpy.ndarray or tensorflow.Dataset. Training image dataset x.
+                The shape of the data should be 3 or 4 dimensional, the last
+                dimension of which should be channel dimension.
```
Comment (fchollet) [inline autokeras/tasks/image.py]: This is rather unclear. Please provide clarification on what the expect input format is. Intuitively one would expect `(samples, width, height, channels)`, so it's not clear where the 3D comes from.

File: autokeras/tasks/image.py [hunk 8]
```diff
@@ -221,3 +221,107 @@ def fit(self,
...
+            y: numpy.ndarray or tensorflow.Dataset. Training image data
+                set y. It should be a tensor with the same shape of x. Each
```
Comment (fchollet) [inline autokeras/tasks/image.py]: It's not the same shape since the number of channels will be different. It's the same *height* and *width*.

**Comment by haifeng-jin** [comment]
@ECEMACHINE Would you please help address Francois's comments?
Note some docstring problems may also exist in the docstrings of other classes.
Please check ImageClassifier, ImageRegressor, ImageInput for similar problems.

Thank you!

File: autokeras/nodes.py [hunk 9]
```diff
@@ -25,7 +25,8 @@ class ImageInput(Input):
     """Input node for image data.
 
     The input data should be numpy.ndarray or tf.data.Dataset. The shape of the data
-    should be 3 or 4 dimensional, the last dimension of which should be channel
+    should be 3 or 4 dimensional, the first three dimensions should be samples, width
+    and height of the image dataset,the last dimension of which should be channel
     dimension.
```
Comment (haifeng-jin) [inline autokeras/nodes.py]: Change it to "should be (samples, width, height) or (samples, width, height, channels)."
Please also follow this style to change the rest.

File: autokeras/tasks/image.py [hunk 10]
```diff
@@ -23,7 +23,8 @@ class ImageClassifier(SupervisedImagePipeline):
     """AutoKeras image classification class.
 
     # Arguments
-        num_classes: Int. Defaults to None. If None, it will infer from the data.
+        num_classes: Int. Defaults to None. If None, it will be inferred from the
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: I just realized there may be more of this grammar error. Would you please use a global search to find them all and change them?

File: autokeras/tasks/image.py [hunk 11]
```diff
@@ -221,3 +224,110 @@ def fit(self,
...
+            overwrite=overwrite,
+            seed=seed)
+
+    def fit(self,
+            x=None,
+            y=None,
+            epochs=None,
+            callbacks=None,
+            validation_split=0.2,
+            validation_data=None,
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: We need type annotations to these arguments as well.

File: autokeras/tasks/image.py [hunk 12]
```diff
@@ -145,15 +147,15 @@ class ImageRegressor(SupervisedImagePipeline):
...
-                 metrics=None,
-                 name='image_regressor',
-                 max_trials=100,
-                 directory=None,
-                 objective='val_loss',
-                 overwrite=True,
-                 seed=None):
+                 output_dim: Optional[int] = None,
+                 loss: str = 'mean_squared_error',
+                 metrics: Optional[List[Union[str, Callable]]] = None,
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: Please follow
https://github.com/keras-team/autokeras/blob/7cade0d7853f25d766f35527df14cbd4fdfedb74/autokeras/tasks/text.py#L47
https://github.com/keras-team/autokeras/blob/7cade0d7853f25d766f35527df14cbd4fdfedb74/autokeras/tasks/text.py#L48
to add the type for loss and metrics.  And for the ImageClassifier and ImageSegmenter, too. The original typing for ImageClassifier may not be correct.

File: autokeras/tasks/image.py [hunk 13]
```diff
@@ -167,22 +169,22 @@ def __init__(self,
...
 
     def fit(self,
-            x=None,
-            y=None,
-            epochs=None,
-            callbacks=None,
-            validation_split=0.2,
-            validation_data=None,
+            x: Any = None,
+            y: Any = None,
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: They should be `Union[np.ndarray, tf.data.Dataset]`.

File: autokeras/tasks/image.py [hunk 14]
```diff
@@ -167,22 +169,22 @@ def __init__(self,
...
+            epochs: Optional[int] = None,
+            callbacks: Optional[List[Union[str, Callable]]] = None,
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: It should be `Optional[List[tf.keras.callbacks.Callback]]`.

File: autokeras/tasks/image.py [hunk 15]
```diff
@@ -145,15 +150,15 @@ class ImageRegressor(SupervisedImagePipeline):
...
-                 loss='mean_squared_error',
-                 metrics=None,
-                 name='image_regressor',
-                 max_trials=100,
-                 directory=None,
-                 objective='val_loss',
-                 overwrite=True,
-                 seed=None):
+                 output_dim: Optional[int] = None,
+                 loss: utils.AcceptableLoss = None,
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: The default value should be `mean_squared_error`.

File: autokeras/tasks/image.py [hunk 16]
```diff
@@ -67,22 +72,22 @@ def __init__(self,
             seed=seed)
 
     def fit(self,
-            x=None,
-            y=None,
-            epochs=None,
-            callbacks=None,
-            validation_split=0.2,
-            validation_data=None,
+            x: Union[np.ndarray, tf.data.Dataset] = None,
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: Change to `x: Optional[Union[np.ndarray, tf.data.Dataset]] = None,`

File: autokeras/tasks/image.py [hunk 17]
```diff
@@ -67,22 +72,22 @@ def __init__(self,
...
+            y: Union[np.ndarray, tf.data.Dataset] = None,
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: This one should be optional, too.

File: autokeras/tasks/image.py [hunk 18]
```diff
@@ -221,3 +226,109 @@ def fit(self,
...
+                 num_classes: Optional[int] = None,
+                 multi_label: bool = False,
+                 loss: utils.AcceptableLoss = None,
+                 metrics: utils.AcceptableMetrics = None,
+                 name: str = 'image_classifier',
+                 max_trials: int = 100,
+                 directory: Union[str, Path, None] = None,
+                 objective: str = 'val_loss',
+                 overwrite: bool = True,
+                 seed: Optional[int] = None,):
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: remove the `,`

File: autokeras/tasks/image.py [hunk 19]
```diff
@@ -67,22 +72,22 @@ def __init__(self,
...
-            epochs=None,
-            callbacks=None,
-            validation_split=0.2,
-            validation_data=None,
+            x: Union[np.ndarray, tf.data.Dataset] = None,
+            y: Union[np.ndarray, tf.data.Dataset] = None,
+            epochs: Optional[int] = None,
+            callbacks: Optional[List[tf.keras.callbacks.Callback]] = None,
+            validation_split: Optional[float] = 0.2,
+            validation_data: Any = None,
```
Comment (haifeng-jin) [inline autokeras/tasks/image.py]: Change to `validation_data: Optional[Union[np.ndarray, tf.data.Dataset]] = None,`
